### PR TITLE
chore: move genesis block builder to chain-spec crate.

### PR DESCRIPTION
**Pull request by yjhmelody**
The genesis block builder is very useful. It should not be defined in `sc-service`.

The logic is better in sc-block-builder. And it's useful to reduce deps when import genesis block builder logic.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2421129

**Comment by bkchr** [comment]
> The genesis block builder is very useful. It should not be defined in `sc-service`.

For what else are you using it?

**Comment by yjhmelody** [comment]
> > The genesis block builder is very useful. It should not be defined in `sc-service`.
> 
> For what else are you using it?

I need to build genesis block for some init checkings in local.

**Review by bkchr — APPROVED** [review]

File: client/block-builder/src/lib.rs [hunk 1]
```diff
@@ -40,9 +45,42 @@ use sp_runtime::{
...
 };
 
+use sc_client_api::backend;
+use sc_executor::RuntimeVersionOf;
 pub use sp_block_builder::BlockBuilder as BlockBuilderApi;
+use sp_core::storage::{well_known_keys, StateVersion, Storage};
 
-use sc_client_api::backend;
+/// Return the state version given the genesis storage and executor.
+pub fn resolve_state_version_from_wasm<E>(
```
Comment (dmitry-markin) [inline client/block-builder/src/lib.rs:54]: The function is duplicated in [`client/service/src/client/client.rs:1166`](https://github.com/paritytech/substrate/blob/e06586108544f39cb71d365552b80b04daa17fd9/client/service/src/client/client.rs#L1166)

Should we keep it only in one place?

File: client/block-builder/src/lib.rs [hunk 2]
```diff
@@ -40,9 +45,42 @@ use sp_runtime::{
...
+			code_fetcher: &code_fetcher,
+			heap_pages: None,
+			hash: {
+				use std::hash::{Hash, Hasher};
+				let mut state = DefaultHasher::new();
+				wasm.hash(&mut state);
+				state.finish().to_le_bytes().to_vec()
+			},
+		};
+		let runtime_version = RuntimeVersionOf::runtime_version(executor, &mut ext, &runtime_code)
```
Comment (koute) [inline client/block-builder/src/lib.rs:75]: Have you considered using [ReadRuntimeVersion](https://paritytech.github.io/substrate/master/sp_core/traits/trait.ReadRuntimeVersion.html) trait + a manual decode instead? I think that could probably work? And it wouldn't require all of the fiddling with `WrappedRuntimeCode` and `RuntimeCode`.

File: client/block-builder/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply client/block-builder/src/lib.rs:54]: Good point and yes!

**Review by bkchr — CHANGES_REQUESTED** [review]
I just realized that I wanted to have it put into some different crate. I think `sc-chain-spec` is better suited for this. Block builder is not really related to what you are doing here.

File: client/block-builder/src/lib.rs (hunk 1 above)
Comment (yjhmelody) [reply client/block-builder/src/lib.rs:54]: Yeah, I just copy this function. Forget to delete this one.

**Comment by yjhmelody** [comment]
Hi @bkchr I found some inconsistecy about state version here:
- Genesis block always use V0 to construct tx. Should it be config?
- Genesis block in test-utils use V1 to construct trie. But use V0 to construct tx

**Comment by yjhmelody** [comment]
> I just realized that I wanted to have it put into some different crate. I think `sc-chain-spec` is better suited for this. Block builder is not really related to what you are doing here.

It is done.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2438914

File: client/chain-spec/src/genesis.rs [hunk 3]
```diff
@@ -93,12 +129,11 @@ impl<Block: BlockT, B: Backend<Block>, E: RuntimeVersionOf> BuildGenesisBlock<Bl
...
 		let Self { genesis_storage, commit_genesis_state, backend, executor, _phantom } = self;
 
-		let genesis_state_version =
-			crate::resolve_state_version_from_wasm(&genesis_storage, &executor)?;
+		let genesis_state_version = resolve_state_version_from_wasm(&genesis_storage, &executor)?;
 		let mut op = backend.begin_operation()?;
 		let state_root =
 			op.set_genesis_state(genesis_storage, commit_genesis_state, genesis_state_version)?;
-		let genesis_block = construct_genesis_block::<Block>(state_root);
+		let genesis_block = construct_genesis_block::<Block>(state_root, StateVersion::V0);
```
Comment (bkchr) [inline client/chain-spec/src/genesis.rs:136]:
```suggestion
		let genesis_block = construct_genesis_block::<Block>(state_root, genesis_state_version);
```

File: test-utils/runtime/client/src/lib.rs [hunk 4]
```diff
@@ -148,7 +150,7 @@ impl substrate_test_client::GenesisInit for GenesisParameters {
 				storage.top.clone().into_iter().chain(child_roots).collect(),
 				sp_runtime::StateVersion::V1,
 			);
-		let block: runtime::Block = client::genesis::construct_genesis_block(state_root);
+		let block: runtime::Block = construct_genesis_block(state_root, StateVersion::V0);
```
Comment (bkchr) [inline test-utils/runtime/client/src/lib.rs:153]:
```suggestion
		let block: runtime::Block = construct_genesis_block(state_root, StateVersion::V1);
```

Above it uses `V1` everywhere.

File: test-utils/runtime/src/genesismap.rs [hunk 5]
```diff
@@ -106,7 +107,7 @@ pub fn insert_genesis_block(storage: &mut Storage) -> sp_core::hash::H256 {
 		storage.top.clone().into_iter().collect(),
 		sp_runtime::StateVersion::V1,
 	);
-	let block: crate::Block = genesis::construct_genesis_block(state_root);
+	let block: crate::Block = construct_genesis_block(state_root, StateVersion::V0);
```
Comment (bkchr) [inline test-utils/runtime/src/genesismap.rs:110]:
```suggestion
	let block: crate::Block = construct_genesis_block(state_root, StateVersion::V1);
```

**Review by bkchr — APPROVED** [review]
Ty!

**Comment by bkchr** [comment]
> * Genesis block always use V0 to construct tx. Should it be config?

Ty for fixing this! This was an oversight!

**Comment by yjhmelody** [comment]
> > * Genesis block always use V0 to construct tx. Should it be config?
> 
> Ty for fixing this! This was an oversight!

Fine, Could you edit the PR title for this fix.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2442614

**Review by dmitry-markin — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2443733

**Review by melekes — APPROVED** [review]
👍

**Comment by bkchr** [comment]
@yjhmelody could you please fix the warnings?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2450835

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
> The CI pipeline was cancelled due to failure one of the required jobs. **Job name**: test-linux-stable-int **Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2450835

@yjhmelody still the same warning as before.

**Pull request merged**